### PR TITLE
Typical values now in all sundials solvers

### PR DESCRIPTION
**Pull request by hsitaram**
1) setting typical values and tolerances are now part of reactorbase, so arkode and cvode can access it
2) fixed bug with arkode ordering for 1darray. Now CYorder is only used when GPUs are used
3) I also did some reorg where the userdatas for cvode and arkode live in reactorutils 

performance using typvals on Eagle CPU/GPU:

for  ncells = 2 256 2, max_grid_size  = 512, ode.dt  = 1.e-04, ode.ndt = 100, ode.rtol = 1e-6, ode.atol = 1e-5
cvode.solve_type = GMRES  
ode.use_erkstep    = 1
ode.typ_vals = 1.e-7 3.e-9 1.e-8 3.e-5 7.e-8 1.e-6 1.e-7 5.e-9 1.e-9 9.e-7 6.e-6 1.e-6 1.7e-8 1.9e-8 7.4e-7 6.7e-9 3.5e-7 1.6e-8 1.3e-8 1.e-4 1.8e-33 2300`

| Solver  | CPU | GPU|
| ------- | -----|-----|
| Erkstep| 219 sec |4.93 sec|
| Cvode  | 29.2 sec  |6.83 sec|

Not sure why Erkstep is super slow on CPU

### CODE REVIEW

File: Reactions/ReactorArkode.cpp [hunk 1]
```diff
@@ -210,15 +212,21 @@ ReactorArkode::react(
   if (use_erkstep == 0) {
     arkode_mem = ARKStepCreate(cF_RHS, nullptr, time, y);
     ARKStepSetUserData(arkode_mem, static_cast<void*>(user_data));
-    ARKStepSStolerances(arkode_mem, relTol, absTol);
-    ARKStepResStolerance(arkode_mem, absTol);
+    setSundialsSolverTols(
+      arkode_mem, user_data->ncells, user_data->verbose, relTol, absTol,
+      "arkstep");
+    // ARKStepSStolerances(arkode_mem, relTol, absTol);
```
Comment (marchdf) [inline Reactions/ReactorArkode.cpp]: remove old code.

File: Reactions/ReactorArkode.cpp [hunk 2]
```diff
@@ -210,15 +212,21 @@ ReactorArkode::react(
...
     ARKStepSetAdaptivityMethod(arkode_mem, rk_controller, 1, 0, nullptr);
     ARKStepEvolve(arkode_mem, time_out, y, &time_init, ARK_NORMAL);
   } else {
     arkode_mem = ERKStepCreate(cF_RHS, time, y);
     ERKStepSetUserData(arkode_mem, static_cast<void*>(user_data));
-    ERKStepSStolerances(arkode_mem, relTol, absTol);
+    setSundialsSolverTols(
+      arkode_mem, user_data->ncells, user_data->verbose, relTol, absTol,
+      "erkstep");
+    // ERKStepSStolerances(arkode_mem, relTol, absTol);
```
Comment (marchdf) [inline Reactions/ReactorArkode.cpp]: remove old code

File: Reactions/ReactorArkode.cpp [hunk 3]
```diff
@@ -358,13 +368,19 @@ ReactorArkode::react(
   if (use_erkstep == 0) {
     arkode_mem = ARKStepCreate(cF_RHS, nullptr, time, y);
     ARKStepSetUserData(arkode_mem, static_cast<void*>(user_data));
-    ARKStepSStolerances(arkode_mem, relTol, absTol);
-    ARKStepResStolerance(arkode_mem, absTol);
+    setSundialsSolverTols(
+      arkode_mem, user_data->ncells, user_data->verbose, relTol, absTol,
+      "arkstep");
+    // ARKStepSStolerances(arkode_mem, relTol, absTol);
```
Comment (marchdf) [inline Reactions/ReactorArkode.cpp]: remove old code

File: Reactions/ReactorArkode.cpp [hunk 4]
```diff
@@ -358,13 +368,19 @@ ReactorArkode::react(
...
+    // ARKStepResStolerance(arkode_mem, absTol);
     ARKStepEvolve(arkode_mem, time_out, y, &time_init, ARK_NORMAL);
   } else {
     arkode_mem = ERKStepCreate(cF_RHS, time, y);
     ERKStepSetUserData(arkode_mem, static_cast<void*>(user_data));
-    ERKStepSStolerances(arkode_mem, relTol, absTol);
+    setSundialsSolverTols(
+      arkode_mem, user_data->ncells, user_data->verbose, relTol, absTol,
+      "erkstep");
+    // ERKStepSStolerances(arkode_mem, relTol, absTol);
```
Comment (marchdf) [inline Reactions/ReactorArkode.cpp]: remove old code

File: Reactions/ReactorBase.H [hunk 5]
```diff
@@ -15,6 +15,8 @@ namespace reactions {
 class ReactorBase : public Factory<ReactorBase>
 {
 public:
+  amrex::Array<double, NUM_SPECIES + 1> typVals;
```
Comment (marchdf) [inline Reactions/ReactorBase.H]: Can you make this protected? Also, don't use double but `amrex::Real`. And then follow the naming convention for the reactor class members: `m_typ_vals`. And initialize it to zero or whatever makes sense.

File: Reactions/ReactorBase.H [hunk 6]
```diff
@@ -78,7 +80,15 @@ public:
     long int* FCunt,
     amrex::Real dt) = 0;
 
-  virtual void SetTypValsODE(const std::vector<amrex::Real>& ExtTypVals) = 0;
+  void SetTypValsODE(const std::vector<amrex::Real>& ExtTypVals);
```
Comment (marchdf) [inline Reactions/ReactorBase.H]: rename to `set_typ_vals_ode`

File: Reactions/ReactorBase.H [hunk 7]
```diff
@@ -78,7 +80,15 @@ public:
...
+
+  void setSundialsSolverTols(
```
Comment (marchdf) [inline Reactions/ReactorBase.H]: rename to `set_sundials_solver_tols`

File: Reactions/ReactorBase.cpp [hunk 8]
```diff
@@ -3,6 +3,117 @@
 namespace pele {
 namespace physics {
 namespace reactions {
+void
+ReactorBase::SetTypValsODE(const std::vector<amrex::Real>& ExtTypVals)
+{
+  int size_ETV = ExtTypVals.size();
```
Comment (marchdf) [inline Reactions/ReactorBase.cpp]: `const` this

File: Reactions/ReactorBase.cpp [hunk 9]
```diff
@@ -3,6 +3,117 @@
...
+  int omp_thread = 0;
+
+#ifdef _OPENMP
+  omp_thread = omp_get_thread_num();
+#endif
+
+  for (int i = 0; i < size_ETV - 1; i++) {
+    typVals[i] = ExtTypVals[i];
+  }
+  typVals[size_ETV - 1] = ExtTypVals[size_ETV - 1];
```
Comment (marchdf) [inline Reactions/ReactorBase.cpp]: why this? Just make the for loop go to `i < size_ETV`  and remove this line?

File: Reactions/ReactorBase.cpp [hunk 10]
```diff
@@ -3,6 +3,117 @@
...
+
+  if (omp_thread == 0) {
+    amrex::Print() << "Set the typVals in PelePhysics: \n  ";
+    for (int i = 0; i < size_ETV - 1; i++) {
+      amrex::Print() << kname[i] << ":" << typVals[i] << "  ";
+    }
+    amrex::Print() << "Temp:" << typVals[size_ETV - 1] << " \n";
```
Comment (marchdf) [inline Reactions/ReactorBase.cpp]: remove `\n` and use `std::endl`

File: Reactions/ReactorBase.cpp [hunk 11]
```diff
@@ -3,6 +3,117 @@
...
+  amrex::Real relTol,
+  amrex::Real absTol,
+  std::string solvername)
+{
+  int omp_thread = 0;
+#ifdef AMREX_USE_OMP
+  omp_thread = omp_get_thread_num();
+#endif
+
+  int neq_tot = (NUM_SPECIES + 1) * ncells;
```
Comment (marchdf) [inline Reactions/ReactorBase.cpp]: `const` this

File: Reactions/ReactorBase.cpp [hunk 12]
```diff
@@ -3,6 +3,117 @@
...
+#endif
+
+  if (typVals[0] > 0.0) {
+    if ((verbose > 0) && (omp_thread == 0)) {
+      amrex::Print() << " Setting " << solvername
+                     << " tolerances with TypVals rtol = " << relTol
+                     << " atolfact = " << absTol << " in PelePhysics \n";
+    }
+    for (int i = 0; i < ncells; i++) {
+      int offset = i * (NUM_SPECIES + 1);
```
Comment (marchdf) [inline Reactions/ReactorBase.cpp]: `const` this

File: Reactions/ReactorBase.cpp [hunk 13]
```diff
@@ -3,6 +3,117 @@
...
+  if (omp_thread == 0) {
+    amrex::Print() << "Set the typVals in PelePhysics: \n  ";
+    for (int i = 0; i < size_ETV - 1; i++) {
+      amrex::Print() << kname[i] << ":" << typVals[i] << "  ";
+    }
+    amrex::Print() << "Temp:" << typVals[size_ETV - 1] << " \n";
+  }
+}
+void
+ReactorBase::setSundialsSolverTols(
```
Comment (marchdf) [inline Reactions/ReactorBase.cpp]: `const` the argument list so that it is clear what is modified and what is not

File: Reactions/ReactorUtils.H [hunk 14]
```diff
@@ -6,10 +6,138 @@
...
+#include <nvector/nvector_cuda.h>
+#include <sunmatrix/sunmatrix_cusparse.h>
+#include <sunlinsol/sunlinsol_cusolversp_batchqr.h>
+#endif
 
 namespace pele {
 namespace physics {
 namespace reactions {
+
+struct CVODEUserData
```
Comment (marchdf) [inline Reactions/ReactorUtils.H]: I dont understand why this was moved here. Can you explain? I kinda liked having arkode and cvode specific stuff in their own specific files.

File: Reactions/ReactorUtils.H [hunk 15]
```diff
@@ -6,10 +6,138 @@
 #include "mechanism.H"
 #include "PelePhysics.H"
 #include "ReactorTypes.H"
+#include <cvode/cvode.h>
```
Comment (marchdf) [inline Reactions/ReactorUtils.H]: Why was this moved here?

File: Reactions/ReactorArkode.cpp (hunk 1 above)
Comment (hsitaram) [reply Reactions/ReactorArkode.cpp]: done

File: Reactions/ReactorArkode.cpp (hunk 2 above)
Comment (hsitaram) [reply Reactions/ReactorArkode.cpp]: done

File: Reactions/ReactorBase.H (hunk 5 above)
Comment (hsitaram) [reply Reactions/ReactorBase.H]: have done this. There are some other variables that dont have the m_ in Arkode and Cvode although.

File: Reactions/ReactorBase.H (hunk 6 above)
Comment (hsitaram) [reply Reactions/ReactorBase.H]: I did this although now the application has to make this change as well. @esclapez - you may have to change LM

File: Reactions/ReactorBase.H (hunk 7 above)
Comment (hsitaram) [reply Reactions/ReactorBase.H]: done

File: Reactions/ReactorBase.cpp (hunk 8 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp]: done

File: Reactions/ReactorArkode.cpp (hunk 3 above)
Comment (hsitaram) [reply Reactions/ReactorArkode.cpp]: done

File: Reactions/ReactorArkode.cpp (hunk 4 above)
Comment (hsitaram) [reply Reactions/ReactorArkode.cpp]: done

File: Reactions/ReactorBase.cpp (hunk 9 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp]: fixed this. I copied over what @esclapez had initially.

File: Reactions/ReactorBase.cpp (hunk 10 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp]: done

File: Reactions/ReactorBase.cpp (hunk 11 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp]: done

File: Reactions/ReactorBase.cpp (hunk 12 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp]: done

File: Reactions/ReactorBase.cpp (hunk 13 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp]: done

File: Reactions/ReactorUtils.H (hunk 14 above)
Comment (hsitaram) [reply Reactions/ReactorUtils.H]: I reverted it back. Previously, cvodeuserdata lived in cvodeutils and arkodeuserdata was in arkode.h. cvodeuserdata should have been cvode.h but it created some conflicts. So I moved both to a common place. I dont know what the "pretty" solution is. So brought it back to original.

File: Reactions/ReactorUtils.H (hunk 15 above)
Comment (hsitaram) [reply Reactions/ReactorUtils.H]: reverted back.

**Comment by hsitaram** [comment]
I also did a test on summit with the same inputs.
Erkstep - 5.72 sec, Cvode - 9.9 sec

File: Reactions/ReactorArkode.H [hunk 16]
```diff
@@ -49,7 +49,11 @@ class ReactorArkode : public ReactorBase::Register<ReactorArkode>
 public:
   static std::string identifier() { return "ReactorArkode"; }
 
+#ifdef AMREX_USE_GPU
```
Comment (jrood-nrel) [inline Reactions/ReactorArkode.H:52]: This was verified to not affect the array4 `react()`?

File: Reactions/ReactorBase.cpp [hunk 17]
```diff
@@ -1,8 +1,140 @@
 #include "ReactorBase.H"
 
+#include <arkode/arkode_arkstep.h>
```
Comment (jrood-nrel) [inline Reactions/ReactorBase.cpp:3]: Should these go in `ReactorBase.H` @marchdf ?

**Review by jrood-nrel — APPROVED** [review]

File: Reactions/ReactorBase.cpp (hunk 17 above)
Comment (hsitaram) [reply Reactions/ReactorBase.cpp:3]: this led me into a rabbit hole of organizing the sundials headers which is not easy.

Comment (jrood-nrel) [reply Reactions/ReactorBase.cpp:3]: Ok no big deal

File: Reactions/ReactorBase.H (hunk 5 above)
Comment (marchdf) [reply Reactions/ReactorBase.H]: yeah... in an ideal world... maybe some day we can do them all.

File: Reactions/ReactorBase.H (hunk 6 above)
Comment (marchdf) [reply Reactions/ReactorBase.H]: hmmm didn't think about that. Maybe that's a bad idea?

File: Reactions/ReactorUtils.H (hunk 14 above)
Comment (marchdf) [reply Reactions/ReactorUtils.H]: Oh I see.... argh yeah I remember trying that...

File: Reactions/ReactorBase.cpp (hunk 17 above)
Comment (marchdf) [reply Reactions/ReactorBase.cpp:3]: yeah it's a pain... I tried to have the headers be included in the files they are actually used. But it's hard...

**Review by marchdf — APPROVED** [review]

**Comment by marchdf** [comment]
Thanks for doing this @hariswaran!

File: Reactions/ReactorBase.H (hunk 6 above)
Comment (esclapez) [reply Reactions/ReactorBase.H]: We don't really have naming convention, but I like no capital on the first letter and capital on each of the word. But this is neither ! We'll need to update LM if this changes, but it's not really a problem.

Comment (marchdf) [reply Reactions/ReactorBase.H]: We have no style guide! What a mess haha. I've been using lowercase and `_` for the reactor classes member and functions so that's the only reason I suggested the change...

File: Reactions/ReactorArkode.H (hunk 16 above)
Comment (hsitaram) [reply Reactions/ReactorArkode.H:52]: Yes its verified, for CPU there doesnt seem to be any appreciable difference in performance.

**Review by esclapez — APPROVED** [review]
